cupy/cuda/memory.pyx

```diff
@@ -486,29 +487,60 @@ cdef class SingleDeviceMemoryPool:
     cpdef Py_ssize_t _round_size(self, Py_ssize_t size):
         """Round up the memory size to fit memory alignment of cudaMalloc."""
         unit = self._allocation_unit_size
-        return (((size + unit - 1) // unit) * unit)
+        return ((size + unit - 1) // unit) * unit
 
     cpdef Py_ssize_t _bin_index_from_size(self, Py_ssize_t size):
         """Get appropriate bins (_free) index from the memory size"""
         unit = self._allocation_unit_size
         return (size - 1) // unit
 
-    cpdef void _grow_free_if_necessary(self, Py_ssize_t size):
+    cpdef _grow_free_if_necessary(self, Py_ssize_t size):
         """Extend bins (_free) size if necessary"""
-        current_size = len(self._free)
-        if current_size >= size:
-            return
-        growth_size = size - current_size
-        growth = [set() for i in range(growth_size)]
-        self._free.extend(growth)
+        if self._index_vector.size() < size:
+            self._index_vector.resize(size, -1)
+
+    cpdef _append_to_free_list(self, Py_ssize_t size, chunk):
+        cdef int index, free_index
+        cdef set free_list
+        index = self._bin_index_from_size(size)
+        try:
+            rlock.lock_fastrlock(self._free_lock, -1, True)
```

Comment:
This line should be moved above the try statement. If `lock_fastrlock` raises, `rlock.unlock_fastrlock` is called unexpectedly. `rlock.lock_fastrlock` may not raise, but moving it above does not harm.

Reply:
FYI: it does not raise unless nonblocking is specified.